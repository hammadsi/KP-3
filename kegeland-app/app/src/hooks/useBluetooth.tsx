import {useEffect} from 'react';
import {NativeEventEmitter, NativeModules, Platform} from 'react-native';
import BleManager from 'react-native-ble-manager';
import {forEach} from 'lodash';

import {
  deviceDisconnected,
  setError,
  setReady,
  stopDeviceScan,
  updateBattery,
} from '~state/ducks/bluetooth/bluetooth.reducer';
import {ANDROID_PERMISSIONS} from '~constants/permissions';
import checkAndroidPermission from '~utils/checkAndroidPermission';
import {connectDevice} from '~state/ducks/bluetooth/bluetooth.actions';
import {PeripheralNotification} from '~constants/bluetooth/interface';

import useAppDispatch from './useAppDispatch';
import useAppSelector from './useAppSelector';

const BleManagerModule = NativeModules.BleManager;
export const bleManagerEmitter = new NativeEventEmitter(BleManagerModule);

/**
 * Custom hook for initiating communication with bluetooth devices
 * @returns true
 */
const useBluetooth = () => {
  const dispatch = useAppDispatch();
  const {connectedDevices, housekeepers} = useAppSelector(
    (state) => state.bluetooth,
  );
  /**
   * Updates state in redux store when scan stopped
   */
  const handleStopScan = () => {
    dispatch(stopDeviceScan());
  };

  /**
   * Reconnects devices stored as 'connected' within the redux state
   */
  const reconnectDevices = () => {
    forEach(connectedDevices, (_, peripheral) => {
      dispatch(connectDevice(peripheral));
    });
  };

  /**
   * Handles battery updates for a device
   * @param param0 the peripheral update information
   * @see {@link PeripheralNotification}
   */
  const handlePeripheralUpdate = ({
    peripheral,
    service,
    value,
  }: PeripheralNotification) => {
    if (service in housekeepers) {
      dispatch(updateBattery({peripheral, value}));
    }
  };

  /**
   * Handles disconnected peripherals
   * @param param0 the peripheral id
   */
  const handleDisconnectedPeripheral = ({peripheral}: {peripheral: string}) => {
    dispatch(deviceDisconnected(peripheral));
  };

  /**
   * Verifies necessary permissions
   */
  const verifyPermissions = async () => {
    if (Platform.OS === 'android' && Platform.Version >= 23) {
      for (const permission of ANDROID_PERMISSIONS) {
        await checkAndroidPermission(permission);
      }
    }
  };

  useEffect(() => {
    bleManagerEmitter.addListener(
      'BleManagerDidUpdateValueForCharacteristic',
      (data: PeripheralNotification) => handlePeripheralUpdate(data),
    );

    return () => {
      bleManagerEmitter.removeAllListeners(
        'BleManagerDidUpdateValueForCharacteristic',
      );
    };
  }, [housekeepers]);

  useEffect(() => {
    BleManager.start({showAlert: false}).catch((err) => {
      if (err instanceof Error) {
        dispatch(setError(err.message));
      } else {
        dispatch(setError(err));
      }
    });

    bleManagerEmitter.addListener('BleManagerStopScan', () => handleStopScan());
    bleManagerEmitter.addListener(
      'BleManagerDisconnectPeripheral',
      (data: {peripheral: string}) => handleDisconnectedPeripheral(data),
    );

    // Verify the app's permissions
    verifyPermissions().then(() => dispatch(setReady()));

    // Reconnect devices from state
    reconnectDevices();

    // Remove listeners when the component is unmounted
    return () => {
      bleManagerEmitter.removeAllListeners('BleManagerStopScan');
      bleManagerEmitter.removeAllListeners('BleManagerDisconnectPeripheral');
      bleManagerEmitter.removeAllListeners(
        'BleManagerDidUpdateValueForCharacteristic',
      );
    };
  }, []);

  return true;
};

export default useBluetooth;
