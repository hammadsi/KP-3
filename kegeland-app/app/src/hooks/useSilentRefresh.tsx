import {useEffect, useState} from 'react';

import {REFRESH_INTERVAL_MS} from '~constants/auth';
import {silentRefresh} from '~state/ducks/auth/auth.actions';

import useAppDispatch from './useAppDispatch';
import useAppSelector from './useAppSelector';

/**
 * Hook for silently refreshing auth-tokens for an authorized user.
 * Will refresh in a given time interval. If the refresh fails, the
 * user will be signed out.
 */
const useSilentRefresh = () => {
  const dispatch = useAppDispatch();
  const {isSignedIn} = useAppSelector((state) => state.auth);
  const [timer, setTimer] = useState<NodeJS.Timer | null>(null);
  useEffect(() => {
    if (isSignedIn) {
      dispatch(silentRefresh());
    }
  }, []);

  useEffect(() => {
    if (isSignedIn) {
      // Start the refresh interval
      const interval = setInterval(() => {
        dispatch(silentRefresh());
      }, REFRESH_INTERVAL_MS);
      setTimer(interval);
    } else {
      if (timer) {
        // Clear interval if the user is not signed in and the interval is initialied
        clearInterval(timer);
        setTimer(null);
      }
    }
    return () => {
      if (timer) {
        // Clear the interval on unmount
        clearInterval(timer);
        setTimer(null);
      }
    };
  }, [isSignedIn]);

  return true;
};

export default useSilentRefresh;
