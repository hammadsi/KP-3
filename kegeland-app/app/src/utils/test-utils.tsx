/* istanbul ignore file */
import React from 'react';
import {NavigationContainer} from '@react-navigation/native';
import {Provider as PaperProvider} from 'react-native-paper';
import {render, RenderOptions} from '@testing-library/react-native';
import {Provider as ReduxProvider} from 'react-redux';

import {store} from '~state/store';
import {lightTheme} from '~constants/theme';
import {initialStore, mockStore} from '~state/ducks/__mocks__/store.mock';

type TestWrapperProps = {
  children: React.ReactNode;
  store?: typeof store;
};

/**
 * Wrapper component with redux state. Used when testing components
 * which requires redux.
 * @param param0 test wrapper props @see {@link TestWrapperProps}
 * @returns
 */
export const TestWrapper: React.FC<TestWrapperProps> = ({
  children,
  ...props
}) => {
  return (
    <ReduxProvider store={props.store || mockStore(initialStore)}>
      <PaperProvider theme={lightTheme}>
        <NavigationContainer theme={lightTheme}>{children}</NavigationContainer>
      </PaperProvider>
    </ReduxProvider>
  );
};

/**
 * Custom test renderer which wraps components with a redux store provider
 * @param ui the component to render
 * @param options renderer options
 * @returns test rendered componnet
 */
export const customRender = <T extends any>(
  ui: React.ReactElement<T>,
  options?: {
    wrapperProps?: {
      store: any;
    };
    renderOptions?: RenderOptions;
  },
) =>
  render(ui, {
    wrapper: (props) => <TestWrapper {...props} {...options?.wrapperProps} />,
    ...options?.renderOptions,
  });

// re-export everything
export * from '@testing-library/react-native';

// override render method
export {customRender as render};
