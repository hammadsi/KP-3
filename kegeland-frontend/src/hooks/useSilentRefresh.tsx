import { useEffect, useState } from 'react';

import { refresh } from '../state/ducks/auth/auth.actions';

import useAppDispatch from './useAppDispatch';
import useAppSelector from './useAppSelector';

export const REFRESH_INTERVAL_MS = 50 * 60 * 1000;
const useSilentRefresh = () => {
  const dispatch = useAppDispatch();
  const { isSignedIn } = useAppSelector((state) => state.auth);
  const [timer, setTimer] = useState<NodeJS.Timer | null>(null);

  useEffect(() => {
    if (isSignedIn) {
      dispatch(refresh());
    }
  }, []);

  useEffect(() => {
    if (isSignedIn) {
      const interval = setInterval(() => {
        dispatch(refresh());
      }, REFRESH_INTERVAL_MS);
      setTimer(interval);
    } else {
      if (timer) {
        clearInterval(timer);
        setTimer(null);
      }
    }
    return () => {
      if (timer) {
        clearInterval(timer);
        setTimer(null);
      }
    };
  }, [isSignedIn]);

  return true;
};

export default useSilentRefresh;
