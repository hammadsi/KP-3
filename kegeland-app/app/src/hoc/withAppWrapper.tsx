import React from 'react';

import useBluetooth from '~hooks/useBluetooth';
import useSilentRefresh from '~hooks/useSilentRefresh';

/**
 * Higher-order component withAppWrapper.
 * Wraps the component with silent refresh- and bluetooth functionality
 * @param Component the component to wrap
 * @see {@link useSilentRefresh}
 * @see {@link useBluetooth}
 */
const withAppWrapper =
  <P extends object>(Component: React.FC<P>): React.FC<P> =>
  (props) => {
    useSilentRefresh();
    useBluetooth();
    return <Component {...props} />;
  };

export default withAppWrapper;
