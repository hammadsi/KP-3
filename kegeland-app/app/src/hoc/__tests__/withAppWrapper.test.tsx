import React from 'react';
import {Text} from 'react-native';

import withAppWrapper from '~hoc/withAppWrapper';
import {render} from '~utils/test-utils';
import * as silentRefreshHook from '~hooks/useSilentRefresh';
import * as bluetoothHook from '~hooks/useBluetooth';

describe('Test withAppWrapper-hoc', () => {
  const MockComponent: React.FC = () => {
    return <Text />;
  };

  it('should render correctly', () => {
    const Component = withAppWrapper(MockComponent);
    const tree = render(<Component />);
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should call correct hooks', () => {
    const silentRefreshHookSpy = jest.spyOn(silentRefreshHook, 'default');
    const bluetoothHookSpy = jest.spyOn(bluetoothHook, 'default');
    const Component = withAppWrapper(MockComponent);
    render(<Component />);
    expect(silentRefreshHookSpy).toBeCalled();
    expect(bluetoothHookSpy).toBeCalled();
  });
});
