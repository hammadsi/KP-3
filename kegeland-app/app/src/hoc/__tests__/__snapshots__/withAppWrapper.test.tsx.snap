// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test withAppWrapper-hoc should render correctly 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Text />
</View>
`;
